src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+        internal HttpHeadersNonValidated(IEnumerable<KeyValuePair<HeaderDescriptor, object>>? headerStore)
+        {
+            _headerStore = headerStore;
+        }
+
+        public IEnumerator<KeyValuePair<string, HeaderStringValues>> GetEnumerator() => _headerStore != null
+            ? new Enumerator(_headerStore)
+            : ((IEnumerable<KeyValuePair<string, HeaderStringValues>>)Array.Empty<KeyValuePair<string, HeaderStringValues>>()).GetEnumerator();
+
+        IEnumerator<KeyValuePair<string, HeaderStringValues>> IEnumerable<KeyValuePair<string, HeaderStringValues>>.GetEnumerator() => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<KeyValuePair<string, HeaderStringValues>>
+        {
+            private readonly IEnumerator<KeyValuePair<HeaderDescriptor, object>> _headerStore;
+
+            internal Enumerator(IEnumerable<KeyValuePair<HeaderDescriptor, object>> headerStore)
+            {
+                _headerStore = headerStore.GetEnumerator();
+                Current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (!_headerStore.MoveNext())
+                {
+                    return false;
+                }
+
+                KeyValuePair<HeaderDescriptor, object> nextPair = _headerStore.Current;
+
+                Current = new KeyValuePair<string, HeaderStringValues>(nextPair.Key.Name, ToStringValues(nextPair.Key, nextPair.Value));
+                return true;
+
+                static HeaderStringValues ToStringValues(HeaderDescriptor descriptor, object value)
+                {
+                    if (value is HttpHeaders.HeaderStoreItemInfo info)
+                    {
+                        if (info.OriginalRawValue != null)
+                        {
+                            return info.OriginalRawValue is string rawString ? new HeaderStringValues(rawString) : new HeaderStringValues((List<string>)info.OriginalRawValue);
+                        }
+
+                        return new HeaderStringValues(HttpHeaders.GetValuesAsStrings(descriptor, info));
+                    }
+
+                    Debug.Assert(value is string);
+                    return new HeaderStringValues((string)value);
+                }
+            }
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
+
+            public void Dispose()
+            {
+                _headerStore.Dispose();
+            }
+
+            object? IEnumerator.Current => Current;
+
+            void IEnumerator.Reset()
+            {
+                _headerStore.Reset();
+            }
+        }
+    }
+
+    public readonly struct HeaderStringValues : IEnumerable<string>
+    {
+        private readonly IEnumerable<string?>? _values;
+        private readonly string? _value;
+
+        public HeaderStringValues(string value)
```

Comment:
This shouldn't be public according to the API.

Reply:
Hmm, as far as I see it, [the approved version](https://github.com/dotnet/runtime/issues/35126#issuecomment-670095983) states this should be `public`.

```C#
    public readonly struct HeaderStringValues : IEnumerable<string>
    {
        public HeaderStringValues(string value);
        public HeaderStringValues(IEnumerable<string> value);
        public Enumerator GetEnumerator();
        public readonly struct Enumerator : IEnumerator<string>
        {
            public bool MoveNext();
            public string Current { get; }
            public void Dispose();
            ... // explicitly implemented interface members
        }
    }
```

Reply:
Sorry, my bad, I wasn't looking properly at the API.